### PR TITLE
Enhance preprocess, add visualize

### PR DESCRIPTION
**Pull request by linhandev**
主要是增强了预处理（添加了针对msd格式数据集的预处理方法，添加对4d数据的支持，简化了prep类创建方式），添加了可视化

目前在msd和新冠数据集上已经可用，矫正体位部分4d数据和sitk打不开的数据还有些问题，方向可能不对但是都能读进来

### CODE REVIEW

File: .pre-commit-config.yaml [hunk 1]
```diff
@@ -8,7 +9,7 @@
         files: \.py$
 
 -   repo: https://github.com/pre-commit/pre-commit-hooks
-    sha: a11d9314b22d8f8c7556443875b731ef05965464
+    rev: a11d9314b22d8f8c7556443875b731ef05965464
```
Comment (shiyutang) [inline .pre-commit-config.yaml:12]: Why is this necessary?

File: paddleseg3d/datasets/preprocess_utils/__init__.py [hunk 2]
```diff
@@ -1,3 +1,5 @@
 from .values import *
 from .geometry import *
 from .uncompress import uncompressor
+from .load_image import sitk_read
+from .file import list_files
```
Comment (shiyutang) [inline paddleseg3d/datasets/preprocess_utils/__init__.py:5]: I have merged list_files with get_image_list in paddleseg3d.utils.utils.py. Please replace the list_files with get_image_list.

File: paddleseg3d/datasets/preprocess_utils/load_image.py [hunk 3]
```diff
@@ -11,9 +11,16 @@
...
 # See the License for the specific language governing permissions and
 # limitations under the License.
+import os
+import math
 
 import numpy as np
 import SimpleITK as sitk
+import pydicom
+
+# DEBUG:
```
Comment (shiyutang) [inline paddleseg3d/datasets/preprocess_utils/load_image.py:21]: Useless annotation can be removed.

File: tools/prepare_msd.py [hunk 4]
```diff
@@ -0,0 +1,173 @@
...
+│   ├── lung_infection
+│   ├── lung_mask
+├── lung_coronavirus_phase0
+│   ├── images
+│   ├── labels
+│   │   ├── coronacases_001.npy
+│   │   ├── ...
+│   │   └── radiopaedia_7_85703_0.npy
+│   ├── train_list.txt
+│   └── val_list.txt
```
Comment (shiyutang) [inline tools/prepare_msd.py:37]: these content should be changed

File: tools/prepare_msd.py [hunk 5]
```diff
@@ -0,0 +1,173 @@
...
+
+        txtname = [
+            osp.join(self.phase_path, "train_list.txt"),
+            osp.join(self.phase_path, "val_list.txt"),
+        ]
+
+        image_files = os.listdir(self.image_path)
+        label_files = [
+            name.replace("_org_covid-19-pneumonia-",
+                         "_").replace("-dcm", "").replace("_org_", "_")
```
Comment (shiyutang) [inline tools/prepare_msd.py:155]: This seem to be unnecessary.

File: tools/prepare_msd.py [hunk 6]
```diff
@@ -0,0 +1,173 @@
...
+
+if __name__ == "__main__":
+    assert (
+        len(sys.argv) == 2
+    ), "Please specify msd task id. \n usage: python tools/prepare_msd.py task_id"
+    prep = Prep_msd(task_id=int(sys.argv[1]))
+    # prep.uncompress_file(num_zipfiles=1)
+    # prep.convert_path()
+    # prep.generate_txt()
+    prep.visualize(alpha=0.1, mode="save", idx=-1)
```
Comment (shiyutang) [inline tools/prepare_msd.py:173]: Have you tested this? What is the result?

File: paddleseg3d/datasets/preprocess_utils/load_image.py [hunk 7]
```diff
@@ -46,4 +53,55 @@ def load_series(mhd_path):
...
+    # 2. change dimensions to z, y, x
+    # s = volume_np.shape
+    # if len(volume_np.shape) == 3:
+    #     if s[0] == s[1]:
+
+    print("volume_np.shape", volume_np.shape)
+    print(sitk_f.GetDirection())
+    print(dim, sitk_f.GetDepth(), sitk_f.GetHeight(), sitk_f.GetWidth())
+    # 3. correct scan orientation, currently only works with 3d image
+    if orient and dim == 3:  # TODO: add 4d orient support
```
Comment (parap1uie-s) [inline paddleseg3d/datasets/preprocess_utils/load_image.py:79]: Check `sitk.DICOMOrient()`

File: tools/prepare.py [hunk 8]
```diff
@@ -88,39 +119,46 @@ def load_save(self,
...
+                    # TODO: nib read 4d series
+                    # TODO: support orient
+                    f_np = nib.load(f).get_fdata(dtype="float32")
+                    f_nps = [f_np.swapaxes(0, 2)]
                 file_suffix = "nii.gz"
             elif "nrrd" in filename:
                 f_np, header = nrrd.read(f)
                 file_suffix = "nrrd"
+                f_nps = [f_np]
             elif "mhd" in filename or "raw" in filename:
```
Comment (parap1uie-s) [inline tools/prepare.py:136]: mhd and raw, 必须同时出现

**Review by parap1uie-s — CHANGES_REQUESTED** [review]
下次不要一次性传这么多内容了

File: paddleseg3d/datasets/preprocess_utils/load_image.py (hunk 7 above)
Comment (linhandev) [reply paddleseg3d/datasets/preprocess_utils/load_image.py:79]: 啊这，搜了一大通没看着有这个api！明天考完试改一下这块

**Comment by linhandev** [comment]
收到，我以后注意

**Comment by linhandev** [comment]
> 下次不要一次性传这么多内容了

收到，我以后注意

File: paddleseg3d/datasets/preprocess_utils/file.py [hunk 9]
```diff
@@ -12,12 +25,15 @@ def list_files(path, filter_suffix=None, filter_key=None, include=None):
     fname = []
     for root, _, f_names in os.walk(path):
         for f in f_names:
+            # skip all hidden files
+            if f.startswith("."):
+                continue
```
Comment (shiyutang) [inline paddleseg3d/datasets/preprocess_utils/file.py:30]: You can add this in the get_image_list

File: tools/prepare_msd.py [hunk 10]
```diff
@@ -0,0 +1,173 @@
...
+            preprocess=[
+                functools.partial(resample, new_shape=[128, 128, 128],
+                                  order=0),
+            ],
+            filter={
+                "filter_suffix": None,
+                "filter_key": None
+            },
+            tag="label",
+        )
```
Comment (shiyutang) [inline tools/prepare_msd.py:142]: load_save module's parameter have been changed  a bit, you can see the repo and change the code accordingly

File: paddleseg3d/datasets/preprocess_utils/load_image.py [hunk 11]
```diff
@@ -46,4 +53,55 @@ def load_series(mhd_path):
...
+
+    # 2. change dimensions to z, y, x
+    # s = volume_np.shape
+    # if len(volume_np.shape) == 3:
+    #     if s[0] == s[1]:
+
+    print("volume_np.shape", volume_np.shape)
+    print(sitk_f.GetDirection())
+    print(dim, sitk_f.GetDepth(), sitk_f.GetHeight(), sitk_f.GetWidth())
+    # 3. correct scan orientation, currently only works with 3d image
```
Comment (shiyutang) [inline paddleseg3d/datasets/preprocess_utils/load_image.py:78]: Please write some annotation, and make sure this work.

File: paddleseg3d/models/unet3d.py [hunk 12]
```diff
@@ -44,189 +45,174 @@ def __init__(self, in_channels, n_classes, base_n_filter=8):
...
+                                     kernel_size=3,
+                                     stride=1,
+                                     padding=1,
+                                     bias_attr=False)
+        self.conv3d_c1_2 = nn.Conv3D(self.base_n_filter,
+                                     self.base_n_filter,
+                                     kernel_size=3,
+                                     stride=1,
+                                     padding=1,
+                                     bias_attr=False)
```
Comment (shiyutang) [inline paddleseg3d/models/unet3d.py:64]: If possible, avoid these changes, it is a bit messy.

File: tools/prepare.py [hunk 13]
```diff
@@ -88,39 +119,46 @@ def load_save(self,
...
+            # load data based on ext, result axis should be in order z, y, x
             if "nii.gz" in filename:
-                f_np = nib.load(f).get_fdata(dtype=np.float32)
+                try:
+                    f_nps = sitk_read(f, split=True)
+                except RuntimeError:
+                    # TODO: nib read 4d series
+                    # TODO: support orient
+                    f_np = nib.load(f).get_fdata(dtype="float32")
+                    f_nps = [f_np.swapaxes(0, 2)]
```
Comment (shiyutang) [inline tools/prepare.py:130]: These code is a bit abnormal, what is the functionality that you want to achieve?

File: tools/prepare.py [hunk 14]
```diff
@@ -88,39 +119,46 @@ def load_save(self,
...
                 file_suffix = "mhd"
             else:
                 raise NotImplementedError
+            file_prefix = filename[:-len(file_suffix) - 1]
 
-            if preprocess is not None:
-                for op in preprocess:
-                    f_np = op(f_np)
+            # plt.imshow(f_nps[0][10])
+            # plt.show()
```
Comment (shiyutang) [inline tools/prepare.py:145]: I have changed the prep.py a bit, this need merge

File: tools/prepare.py [hunk 15]
```diff
@@ -131,8 +169,76 @@ def generate_txt(self):
         raise NotImplementedError
 
     # TODO add data visualize method, such that data can be checked every time after preprocess.
-    def visualize(self):
-        pass
+    def visualize(self, idx=None, alpha=0.8, mode="show"):
```
Comment (shiyutang) [inline tools/prepare.py:172]: what is the effect of the visualization? it is recommended to show it on the PR comment. 
This function can split into modules. Currently, it is a bit too complicate.

**Review by shiyutang — CHANGES_REQUESTED** [review]

**Review by shiyutang — COMMENTED** [review]
There are many messy format changes(no content has changed), this is not encouraged. If possible, please remove it.

File: paddleseg3d/datasets/preprocess_utils/load_image.py (hunk 11 above)
Comment (linhandev) [reply paddleseg3d/datasets/preprocess_utils/load_image.py:78]: will do, all complex functions would have comment

File: paddleseg3d/models/unet3d.py (hunk 12 above)
Comment (linhandev) [reply paddleseg3d/models/unet3d.py:64]: I think this is done by pre-commit. Maybe I have misconfigured something. I thinkg I can disable it and only use formatter on new files I add.

File: tools/prepare.py (hunk 13 above)
Comment (linhandev) [reply tools/prepare.py:130]: sitk wouldn't read scans that lack orthonormal direction cosines in header. I think that matrix is necessary for orientation correction. [parap1uie-s](https://github.com/parap1uie-s) mentioned an alternative way to do orientation correction. I will try to remove this try catch.
